### ISSUE TITLE
libtiledbsoma build fails with libfmt errors when linked to libtiledb 2.27.0-rc1

### ISSUE THREAD

**Issue by jdblischak** [TileDB-Inc/centralized-tiledb-nightlies#26]
This morning I tested the latest libtiledb [2.27.0-rc1](https://github.com/TileDB-Inc/TileDB/releases/tag/2.27.0-rc1) release binary in the centralized nightlies pipeline

* [branch](https://github.com/TileDB-Inc/centralized-tiledb-nightlies/tree/227rc1)
* [commit](https://github.com/TileDB-Inc/centralized-tiledb-nightlies/commit/e1140cd6ab831f965a5d6fabfeee31919fca8422)
* [build](https://github.com/TileDB-Inc/centralized-tiledb-nightlies/actions/runs/11917656965)

The libtiledbsoma build failed with libfmt-related errors.

cc: @johnkerl, @ihnorton 

```sh
[  5%] Building CXX object src/CMakeFiles/TILEDB_SOMA_OBJECTS.dir/soma/managed_query.cc.o
/home/runner/work/centralized-tiledb-nightlies/centralized-tiledb-nightlies/libtiledbsoma/src/soma/managed_query.cc:37:10: fatal error: format: No such file or directory
   37 | #include <format>
      |          ^~~~~~~~
compilation terminated.
gmake[5]: *** [src/CMakeFiles/TILEDB_SOMA_OBJECTS.dir/build.make:93: src/CMakeFiles/TILEDB_SOMA_OBJECTS.dir/soma/managed_query.cc.o] Error 1
gmake[5]: *** Waiting for unfinished jobs....


[ 10%] Building CXX object _deps/catch2-build/src/CMakeFiles/Catch2.dir/catch2/reporters/catch_reporter_compact.cpp.o
/home/runner/work/centralized-tiledb-nightlies/centralized-tiledb-nightlies/libtiledbsoma/src/reindexer/reindexer.cc: In member function ‘void tiledbsoma::IntIndexer::map_locations(const int64_t*, size_t)’:
/home/runner/work/centralized-tiledb-nightlies/centralized-tiledb-nightlies/libtiledbsoma/src/reindexer/reindexer.cc:63:14: error: ‘format’ is not a member of ‘std’; did you mean ‘fmt::v8::format’?
   63 |         std::format("[Re-indexer] Start of Map locations with {} keys", size));
      |              ^~~~~~
In file included from /home/runner/work/centralized-tiledb-nightlies/centralized-tiledb-nightlies/build-libtiledbsoma/externals/install/include/spdlog/fmt/bundled/core.h:3234,
                 from /home/runner/work/centralized-tiledb-nightlies/centralized-tiledb-nightlies/build-libtiledbsoma/externals/install/include/spdlog/fmt/fmt.h:24,
                 from /home/runner/work/centralized-tiledb-nightlies/centralized-tiledb-nightlies/build-libtiledbsoma/externals/install/include/spdlog/common.h:45,
                 from /home/runner/work/centralized-tiledb-nightlies/centralized-tiledb-nightlies/build-libtiledbsoma/externals/install/include/spdlog/spdlog.h:12,
                 from /home/runner/work/centralized-tiledb-nightlies/centralized-tiledb-nightlies/libtiledbsoma/src/utils/logger.h:54,
                 from /home/runner/work/centralized-tiledb-nightlies/centralized-tiledb-nightlies/libtiledbsoma/src/reindexer/reindexer.cc:41:
/home/runner/work/centralized-tiledb-nightlies/centralized-tiledb-nightlies/build-libtiledbsoma/externals/install/include/spdlog/fmt/bundled/format.h:3058:13: note: ‘fmt::v8::format’ declared here
 3058 | inline auto format(const Locale& loc, format_string<T...> fmt, T&&... args)
      |             ^~~~~~
/home/runner/work/centralized-tiledb-nightlies/centralized-tiledb-nightlies/libtiledbsoma/src/reindexer/reindexer.cc:74:20: error: ‘format’ is not a member of ‘std’; did you mean ‘fmt::v8::format’?
   74 |     LOG_DEBUG(std::format("[Re-indexer] khash size = {}", hsize));
      |                    ^~~~~~
In file included from /home/runner/work/centralized-tiledb-nightlies/centralized-tiledb-nightlies/build-libtiledbsoma/externals/install/include/spdlog/fmt/bundled/core.h:3234,
                 from /home/runner/work/centralized-tiledb-nightlies/centralized-tiledb-nightlies/build-libtiledbsoma/externals/install/include/spdlog/fmt/fmt.h:24,
                 from /home/runner/work/centralized-tiledb-nightlies/centralized-tiledb-nightlies/build-libtiledbsoma/externals/install/include/spdlog/common.h:45,
                 from /home/runner/work/centralized-tiledb-nightlies/centralized-tiledb-nightlies/build-libtiledbsoma/externals/install/include/spdlog/spdlog.h:12,
                 from /home/runner/work/centralized-tiledb-nightlies/centralized-tiledb-nightlies/libtiledbsoma/src/utils/logger.h:54,
                 from /home/runner/work/centralized-tiledb-nightlies/centralized-tiledb-nightlies/libtiledbsoma/src/reindexer/reindexer.cc:41:
/home/runner/work/centralized-tiledb-nightlies/centralized-tiledb-nightlies/build-libtiledbsoma/externals/install/include/spdlog/fmt/bundled/format.h:3058:13: note: ‘fmt::v8::format’ declared here
 3058 | inline auto format(const Locale& loc, format_string<T...> fmt, T&&... args)
      |             ^~~~~~
/home/runner/work/centralized-tiledb-nightlies/centralized-tiledb-nightlies/libtiledbsoma/src/reindexer/reindexer.cc:77:14: error: ‘format’ is not a member of ‘std’; did you mean ‘fmt::v8::format’?
   77 |         std::format("[Re-indexer] Thread pool started and hash table created"));
      |              ^~~~~~
In file included from /home/runner/work/centralized-tiledb-nightlies/centralized-tiledb-nightlies/build-libtiledbsoma/externals/install/include/spdlog/fmt/bundled/core.h:3234,
                 from /home/runner/work/centralized-tiledb-nightlies/centralized-tiledb-nightlies/build-libtiledbsoma/externals/install/include/spdlog/fmt/fmt.h:24,
                 from /home/runner/work/centralized-tiledb-nightlies/centralized-tiledb-nightlies/build-libtiledbsoma/externals/install/include/spdlog/common.h:45,
                 from /home/runner/work/centralized-tiledb-nightlies/centralized-tiledb-nightlies/build-libtiledbsoma/externals/install/include/spdlog/spdlog.h:12,
                 from /home/runner/work/centralized-tiledb-nightlies/centralized-tiledb-nightlies/libtiledbsoma/src/utils/logger.h:54,
                 from /home/runner/work/centralized-tiledb-nightlies/centralized-tiledb-nightlies/libtiledbsoma/src/reindexer/reindexer.cc:41:
/home/runner/work/centralized-tiledb-nightlies/centralized-tiledb-nightlies/build-libtiledbsoma/externals/install/include/spdlog/fmt/bundled/format.h:3058:13: note: ‘fmt::v8::format’ declared here
 3058 | inline auto format(const Locale& loc, format_string<T...> fmt, T&&... args)
      |             ^~~~~~
/home/runner/work/centralized-tiledb-nightlies/centralized-tiledb-nightlies/libtiledbsoma/src/reindexer/reindexer.cc: In member function ‘void tiledbsoma::IntIndexer::lookup(const int64_t*, int64_t*, size_t)’:
/home/runner/work/centralized-tiledb-nightlies/centralized-tiledb-nightlies/libtiledbsoma/src/reindexer/reindexer.cc:98:20: error: ‘format’ is not a member of ‘std’; did you mean ‘fmt::v8::format’?
   98 |     LOG_DEBUG(std::format(
      |                    ^~~~~~
In file included from /home/runner/work/centralized-tiledb-nightlies/centralized-tiledb-nightlies/build-libtiledbsoma/externals/install/include/spdlog/fmt/bundled/core.h:3234,
                 from /home/runner/work/centralized-tiledb-nightlies/centralized-tiledb-nightlies/build-libtiledbsoma/externals/install/include/spdlog/fmt/fmt.h:24,
                 from /home/runner/work/centralized-tiledb-nightlies/centralized-tiledb-nightlies/build-libtiledbsoma/externals/install/include/spdlog/common.h:45,
                 from /home/runner/work/centralized-tiledb-nightlies/centralized-tiledb-nightlies/build-libtiledbsoma/externals/install/include/spdlog/spdlog.h:12,
                 from /home/runner/work/centralized-tiledb-nightlies/centralized-tiledb-nightlies/libtiledbsoma/src/utils/logger.h:54,
                 from /home/runner/work/centralized-tiledb-nightlies/centralized-tiledb-nightlies/libtiledbsoma/src/reindexer/reindexer.cc:41:
/home/runner/work/centralized-tiledb-nightlies/centralized-tiledb-nightlies/build-libtiledbsoma/externals/install/include/spdlog/fmt/bundled/format.h:3058:13: note: ‘fmt::v8::format’ declared here
 3058 | inline auto format(const Locale& loc, format_string<T...> fmt, T&&... args)
      |             ^~~~~~
/home/runner/work/centralized-tiledb-nightlies/centralized-tiledb-nightlies/libtiledbsoma/src/reindexer/reindexer.cc:117:24: error: ‘format’ is not a member of ‘std’; did you mean ‘fmt::v8::format’?
  117 |         LOG_DEBUG(std::format(
      |                        ^~~~~~
In file included from /home/runner/work/centralized-tiledb-nightlies/centralized-tiledb-nightlies/build-libtiledbsoma/externals/install/include/spdlog/fmt/bundled/core.h:3234,
                 from /home/runner/work/centralized-tiledb-nightlies/centralized-tiledb-nightlies/build-libtiledbsoma/externals/install/include/spdlog/fmt/fmt.h:24,
                 from /home/runner/work/centralized-tiledb-nightlies/centralized-tiledb-nightlies/build-libtiledbsoma/externals/install/include/spdlog/common.h:45,
                 from /home/runner/work/centralized-tiledb-nightlies/centralized-tiledb-nightlies/build-libtiledbsoma/externals/install/include/spdlog/spdlog.h:12,
                 from /home/runner/work/centralized-tiledb-nightlies/centralized-tiledb-nightlies/libtiledbsoma/src/utils/logger.h:54,
                 from /home/runner/work/centralized-tiledb-nightlies/centralized-tiledb-nightlies/libtiledbsoma/src/reindexer/reindexer.cc:41:
/home/runner/work/centralized-tiledb-nightlies/centralized-tiledb-nightlies/build-libtiledbsoma/externals/install/include/spdlog/fmt/bundled/format.h:3058:13: note: ‘fmt::v8::format’ declared here
 3058 | inline auto format(const Locale& loc, format_string<T...> fmt, T&&... args)
      |             ^~~~~~
/home/runner/work/centralized-tiledb-nightlies/centralized-tiledb-nightlies/libtiledbsoma/src/reindexer/reindexer.cc:134:24: error: ‘format’ is not a member of ‘std’; did you mean ‘fmt::v8::format’?
  134 |         LOG_DEBUG(std::format(
      |                        ^~~~~~
In file included from /home/runner/work/centralized-tiledb-nightlies/centralized-tiledb-nightlies/build-libtiledbsoma/externals/install/include/spdlog/fmt/bundled/core.h:3234,
                 from /home/runner/work/centralized-tiledb-nightlies/centralized-tiledb-nightlies/build-libtiledbsoma/externals/install/include/spdlog/fmt/fmt.h:24,
                 from /home/runner/work/centralized-tiledb-nightlies/centralized-tiledb-nightlies/build-libtiledbsoma/externals/install/include/spdlog/common.h:45,
                 from /home/runner/work/centralized-tiledb-nightlies/centralized-tiledb-nightlies/build-libtiledbsoma/externals/install/include/spdlog/spdlog.h:12,
                 from /home/runner/work/centralized-tiledb-nightlies/centralized-tiledb-nightlies/libtiledbsoma/src/utils/logger.h:54,
                 from /home/runner/work/centralized-tiledb-nightlies/centralized-tiledb-nightlies/libtiledbsoma/src/reindexer/reindexer.cc:41:
/home/runner/work/centralized-tiledb-nightlies/centralized-tiledb-nightlies/build-libtiledbsoma/externals/install/include/spdlog/fmt/bundled/format.h:3058:13: note: ‘fmt::v8::format’ declared here
 3058 | inline auto format(const Locale& loc, format_string<T...> fmt, T&&... args)
      |             ^~~~~~
```

**Comment by jdblischak** [comment]
> As of C++ 20 we can use `std::format` instead. Core has been at C++ 20 for a while; tiledbsoma is at C++ 20 as of 2024-11-19. Tracking issue: https://github.com/single-cell-data/TileDB-SOMA/issues/3154.

@johnkerl Great! Thanks for the context. It's good to know that this is being actively worked on so I can expect it to be in flux.

> Let's check tonight's run.

Tonight's scheduled run will build the main branch of libtiledbsoma against the dev branch of libtiledb. I went ahead and manually triggered another build on my branch that builds the main branch of libtiledbsoma against the release binary for 2.27.0-rc1.

https://github.com/TileDB-Inc/centralized-tiledb-nightlies/actions/runs/11919421932

The original build failure I described above used TileDB-SOMA commit https://github.com/single-cell-data/TileDB-SOMA/commit/f5a3dbc5c7224cb5fef6b71145e811920bc651af (the one that migrated to C++20), and the latest one used https://github.com/single-cell-data/TileDB-SOMA/commit/8b6ab64992c71a5d84ce7b690f7afc379791337b (only a CI update). I should have checked this before I manually triggered the build!

In summary, the build failure was with the C++20 enabled TileDB-SOMA, which in retrospect makes sense, since the error messages mention `std::format` instead of `fmt::format`.

We'll see what happens tonight when it is built against dev libtiledb.

**Comment by jdblischak** [comment]
I rebased my [branch](https://github.com/TileDB-Inc/centralized-tiledb-nightlies/tree/refs/heads/227rc1) onto main to build with a newer GCC (#27). libtiledbsoma successfully [built](https://github.com/TileDB-Inc/centralized-tiledb-nightlies/actions/runs/11920846832) against the release binary for libtiledb 2.27.0-rc1 🎉 Thanks @johnkerl!
